### PR TITLE
fix: Make hydra integration branch compatible with kratos-0.5.4, hydra-1.8.5

### PR DESCRIPTION
**Pull request by ericb-summit**
Also removed all unnecessary deltas over master.

Based on previous commits in hydra-integration branch by:
- Kim Neunert <kneunert@gmail.com>
- Jeremy Bogle <jbogle@apple.com>
- hackerman <3372410+aeneasr@users.noreply.github.com>
- Alexander Konotop <aleksandr.k@mserverone.com>

Other squashed commit messages:
- preliminary ORY Hydra integration (#50)
- fix: redirect issues for hydra integration in domain (not port) based env (#68)
  See #67

## Related issue

Sending PR as discussed with @vinckr on slack

## Proposed changes

This PR makes the hydra-integration branch work with recent version components, kratos-0.5.4, hydra-1.8.5

Other irrelevant changes of hydra-integration over master (formatting, coded moved around, console vs logger) were removed, so next person doing a rebase can do so more easily.

Affected code is formatted using `npm run format`, unaffected code is not.  `npm test` seems fine, not aware of any other tests.  Known issue of the hydra_integration branch,  works on firefox, not on chrome (I may tackle that later); that issue persists.

## Checklist

- [X] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md)
      and signed the CLA.
- [X] I have read the [security policy](../security/policy).
- [X] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added necessary documentation within the code base (if
      appropriate).

## Further comments

We really ought to add some coverage to this.

### CODE REVIEW

**Comment by vinckr** [comment]
Beautiful PR :) 
I will check it out over the weekend!

**Comment by ericb-summit** [comment]
Cool.  FYI, circleci fails because the base (master) is not prettied.  I don't think it a good idea to add unrelated stuff to this PR.

File: contrib/hydra/kratos/oidc.github.jsonnet [hunk 1]
```diff
@@ -0,0 +1,17 @@
+local claims = {
```
Comment (aeneasr) [inline contrib/hydra/kratos/oidc.github.jsonnet:1]: OIDC isn't enabled in the config - can these files be removed?

File: src/index.ts [hunk 2]
```diff
@@ -17,6 +24,45 @@ import * as https from 'https'
 import * as fs from 'fs'
 import protectOathkeeper from './middleware/oathkeeper'
 import protectSimple from './middleware/simple'
+import bodyParser from 'body-parser'
+import csrf from 'csurf'
+import session from 'express-session'
+
+const csrfProtection = csrf({ cookie: true })
+
+const protectOathKeeper = jwt({
```
Comment (aeneasr) [inline src/index.ts:33]: Given that we don't really use this set up any more - should we just remove it? Alternatively, this is handled in: https://github.com/ory/kratos-selfservice-ui-node/blob/master/src/middleware/oathkeeper.ts

File: src/routes/hydra.ts [hunk 3]
```diff
@@ -0,0 +1,355 @@
+import { NextFunction, Request, Response } from 'express'
```
Comment (aeneasr) [inline src/routes/hydra.ts:1]: The diff is a bit difficult - are there code changes from the original one?

**Review by aeneasr — COMMENTED** [review]
Awesome, thank you for your contribution! This looks pretty good and I have some ideas how to improve it further :)

File: contrib/hydra/kratos/oidc.github.jsonnet (hunk 1 above)
Comment (ericb-summit) [reply contrib/hydra/kratos/oidc.github.jsonnet:1]: Agreed, they can.

File: src/routes/hydra.ts (hunk 3 above)
Comment (ericb-summit) [reply src/routes/hydra.ts:1]: Unchanged from hydra-integration branch with the exception of adapting to newer hydra/kratos npm API.

File: src/index.ts (hunk 2 above)
Comment (ericb-summit) [reply src/index.ts]: Also correct, I missed this duplicate.  Done.

**Comment by ericb-summit** [comment]
@aeneasr updated per your comments.

Given these changes have no impact on how the project works if not integrated with hydra, can I suggest that we merge this to master?

This way, as the kratos/hydra APIs are updated, the hydra.ts code will at a minimum compile.

File: src/routes/hydra.ts (hunk 3 above)
Comment (aeneasr) [reply src/routes/hydra.ts:1]: Thank you for the clarification!

File: contrib/hydra/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      - ./pg-init:/docker-entrypoint-initdb.d
+    networks:
+      - intranet
+
+  hydra-migrate:
+    image: oryd/hydra:v1.8.5-alpine
+    depends_on:
+      - postgresd
+    environment:
+      - DSN=postgres://pguser:secret@postgresd:5432/hydra?sslmode=disable
```
Comment (aeneasr) [inline contrib/hydra/docker-compose.yml:25]: We could also use SQLite here and remove the need for pg-init et al?

File: src/config.ts [hunk 5]
```diff
@@ -31,6 +38,8 @@ export default {
   SECURITY_MODE_JWT,
   SECURITY_MODE_STANDALONE,
 
+  cookieSecret: process.env.COOKIE_SECRET || cookieSecret,
```
Comment (aeneasr) [inline src/config.ts:41]: I would move this key under `hydra` and also prefix it `HYDRA_COOKIE_SECRET`

File: src/index.ts [hunk 6]
```diff
@@ -17,6 +22,35 @@ import * as https from 'https'
...
+import bodyParser from 'body-parser'
+import csrf from 'csurf'
+import session from 'express-session'
+
+const csrfProtection = csrf({ cookie: true })
+
+const publicEndpoint = new PublicApi(
+  new Configuration({ basePath: config.kratos.public })
+)
+const protectProxy = (req: Request, res: Response, next: NextFunction) => {
```
Comment (aeneasr) [inline src/index.ts:34]: This code is implemented here https://github.com/ory/kratos-selfservice-ui-node/blob/master/src/middleware/simple.ts - is there a particular reason for not using that instead?

File: src/index.ts [hunk 7]
```diff
@@ -84,11 +135,43 @@ if (process.env.NODE_ENV === 'stub') {
...
+      csrfProtection,
+      hydraPostConsent
+    )
+  }
 }
 
 app.get('/health', (_: Request, res: Response) => res.send('ok'))
 app.get('/debug', debug)
 
+if (config.securityMode === SECURITY_MODE_STANDALONE) {
```
Comment (aeneasr) [inline src/index.ts:161]: Do we need this? We removed it as it no longer had a purpose!

File: src/test.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+const url = new URL(
```
Comment (aeneasr) [inline src/test.js:1]: Does this serve a purpose?

**Review by aeneasr — COMMENTED** [review]
> Given these changes have no impact on how the project works if not integrated with hydra, can I suggest that we merge this to master?
>
> This way, as the kratos/hydra APIs are updated, the hydra.ts code will at a minimum compile.

First of all, thank you for your effort! The reason I am reluctant to merge this into master is that we did not properly audit it and that there are no e2e tests. Otherwise I would be completely open to adopt it in the master branch!

File: src/config.ts (hunk 5 above)
Comment (ericb-summit) [reply src/config.ts:41]: Agreed

File: contrib/hydra/docker-compose.yml (hunk 4 above)
Comment (ericb-summit) [reply contrib/hydra/docker-compose.yml:25]: Yeah, no idea why this was on postgres, figured the original committer had a reason

File: src/index.ts (hunk 7 above)
Comment (ericb-summit) [reply src/index.ts:161]: Based on how the integration was originally implemented, this proxying is required.  The browser does not speak to Kratos directly, it always goes through the selfservice-ui.

File: src/index.ts (hunk 6 above)
Comment (ericb-summit) [reply src/index.ts:34]: Agreed, this is now dead code.  Removed.

File: src/test.js (hunk 8 above)
Comment (ericb-summit) [reply src/test.js:1]: Seems useless.

**Comment by ericb-summit** [comment]
> First of all, thank you for your effort! The reason I am reluctant to merge this into master is that we did not properly audit it and that there are no e2e tests. Otherwise I would be completely open to adopt it in the master branch!

Fair enough.  I changed the target branch back to hydra-integration-2021.

**Comment by ericb-summit** [comment]
BTW, what framework would we use to add tests to this?

**Comment by aeneasr** [comment]
> BTW, what framework would we use to add tests to this?

Basically we check out the repo in ory kratos and start it up together with some databases and kratos itself, and then run cypress tests: https://github.com/ory/kratos/tree/master/test/e2e

I will now look over your changes!

**Comment by aeneasr** [comment]
Oh and could you please run `npm run format` to pass the CI? :)

File: src/index.ts (hunk 7 above)
Comment (aeneasr) [reply src/index.ts:137]: Ok, thank you for the clarification!

**Review by aeneasr — APPROVED** [review]
Awesome 🎉 

Thank you for your contribution!

**Comment by ericb-summit** [comment]
> Oh and could you please run `npm run format` to pass the CI? :)

Actually there were changes the format in `package.json`, that's wht broke CI.  I reverted, now it passes.

**Comment by aeneasr** [comment]
🥳

**Pull request merged**